schemas/README.md

```diff
@@ -1593,6 +2172,548 @@ Raw image data
+
+<table>
+  <tr>
+    <th>field</th>
+    <th>type</th>
+    <th>description</th>
+  </tr>
+<tr>
+<td><code>timestamp</code></td>
+<td>
+
+time
+
+</td>
+<td>
+
+Timestamp of the entity
+
+</td>
+</tr>
+<tr>
+<td><code>frame_id</code></td>
+<td>
+
+string
+
+</td>
+<td>
+
+Frame of reference
+
+</td>
+</tr>
+<tr>
+<td><code>id</code></td>
+<td>
+
+string
+
+</td>
+<td>
+
+Identifier for the entity. A entity will replace any prior entity on the same topic with the same `id`.
+
+</td>
+</tr>
+<tr>
+<td><code>lifetime</code></td>
+<td>
+
+duration
+
+</td>
+<td>
+
+Length of time (relative to `timestamp`) after which the entity should be automatically removed. Zero value indicates the entity should remain visible until it is replaced or deleted.
+
+</td>
+</tr>
+<tr>
+<td><code>frame_locked</code></td>
+<td>
+
+boolean
+
+</td>
+<td>
+
+Whether the entity should keep its location in the fixed frame (false) or follow the frame specified in `frame_id` as it moves relative to the fixed frame (true)
+
+</td>
+</tr>
+<tr>
+<td><code>metadata</code></td>
+<td>
+
+[KeyValuePair](#keyvaluepair)[]
+
+</td>
+<td>
+
+Additional user-provided metadata associated with the entity. Keys must be unique.
+
+</td>
+</tr>
+<tr>
+<td><code>arrows</code></td>
+<td>
+
+[ArrowPrimitive](#arrowprimitive)[]
+
+</td>
+<td>
+
+Arrow primitives
+
+</td>
+</tr>
+<tr>
+<td><code>cubes</code></td>
+<td>
+
+[CubePrimitive](#cubeprimitive)[]
+
+</td>
+<td>
+
+Cube primitives
+
+</td>
+</tr>
+<tr>
+<td><code>spheres</code></td>
+<td>
+
+[SpherePrimitive](#sphereprimitive)[]
+
+</td>
+<td>
+
+Sphere primitives
+
+</td>
+</tr>
+<tr>
+<td><code>cylinders</code></td>
+<td>
+
+[CylinderPrimitive](#cylinderprimitive)[]
+
+</td>
+<td>
+
+Cylinder primitives
+
+</td>
+</tr>
+<tr>
+<td><code>lines</code></td>
+<td>
+
+[LinePrimitive](#lineprimitive)[]
+
+</td>
+<td>
+
+Line primitives
+
+</td>
+</tr>
+<tr>
+<td><code>triangles</code></td>
+<td>
+
+[TriangleListPrimitive](#trianglelistprimitive)[]
+
+</td>
+<td>
+
+Triangle list primitives
+
+</td>
+</tr>
+<tr>
+<td><code>texts</code></td>
+<td>
+
+[TextPrimitive](#textprimitive)[]
+
+</td>
+<td>
+
+Text primitives
+
+</td>
+</tr>
+<tr>
+<td><code>models</code></td>
+<td>
+
+[ModelPrimitive](#modelprimitive)[]
+
+</td>
+<td>
+
+Model primitives
+
+</td>
+</tr>
+</table>
+
+## SceneEntityDeletion
+
+Command to remove previously published entities
+
+<table>
+  <tr>
+    <th>field</th>
+    <th>type</th>
+    <th>description</th>
+  </tr>
+<tr>
+<td><code>timestamp</code></td>
+<td>
+
+time
+
+</td>
+<td>
+
+Timestamp of the deletion. Only matching entities earlier than this timestamp will be deleted.
+
+</td>
+</tr>
+<tr>
+<td><code>type</code></td>
+<td>
+
+[enum SceneEntityDeletionType](#enum-sceneentitydeletiontype)
+
+</td>
+<td>
+
+Type of deletion action to perform
+
+</td>
+</tr>
+<tr>
+<td><code>id</code></td>
```

Comment:
Did you consider accepting an array of `ids` to delete? This would also make it simpler to publish `SceneEntityDeletion` directly on a topic without needing to wrap it in `SceneEntityUpdate`.

Reply:
I didn't think about it. Would you still combine this with the `type` field (or `delete_all` boolean if we rename it that)? Would we remove the `deletions` array from SceneEntityUpdate or would it still be useful enough to be able to delete+add in the same batch?

Reply:
Yes I guess we could potentially remove `deletions` from the `SceneEntityUpdate` and just ask you to publish a deletion followed by addition. If we did that it would then make sense to rename `SceneEntityUpdate` to just `SceneEntities` since it would only include additions.

I'm not sure how common different workflows will be, but I could totally see `delete_all` being commonly used alongside additions (if you want to simply redraw the world). It might be less common to include specific ID deletions alongside additions, although still plausible (if you had some node publishing updates from a tracker).

Net/net I feel like I would lean towards keeping `SceneUpdate` (aka `SceneEntityUpdate`), which includes options to (a) clear scene (delete all), (b) delete one or more entities by IDs, and (c) add one or more entities. In addition, it feels cleaner to _require_ `SceneUpdate` and not support publishing individual scene entity / deletion messages directly on a topic, since that would result in multiple schemas being published on the same topic, which is technically legal but unnecessarily messy.

thoughts?

Reply:
> It might be less common to include specific ID deletions alongside additions, although still plausible (if you had some node publishing updates from a tracker).

Sounds common to me! At least in self driving...

> require SceneUpdate and not support publishing individual scene entity / deletion messages directly on a topic, since that would result in multiple schemas being published on the same topic

Why would that result in multiple schemas on the same topic? You could have zero or more topics that are `SceneUpdate`, and zero or more that are `SceneEntityDeletion` or `SceneEntity`, that doesn't mean a single topic is switching its schema mid-stream.

Reply:
Isn't the `id` field effectively namespaced per topic? If I published `SceneEntity` on topic A and `SceneEntityDeletion` on topic B, I would not expect it to delete any objects from topic A.

Reply:
Oh yeah that's right